Deduplicate hosts added to gateway listeners

maxbischoff opened the pull request:
When building SNI hosts for a server's TLS filter chains hosts are sanitized to remove the `namespace/` prefix if set. To ensure that no duplicates exist after this, only hosts that weren't in the SNI host list are added to the list. When the resulting config contains duplicate hosts envoy rejects it.

Signed-off-by: Maximilian Bischoff <maximilian.bischoff@inovex.de>

Solves #16573

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@maxbischoff: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-unit-tests-master | d1e82be570c6c460a32679628f689d8008db5895 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-unit-tests-master/6415) | `/test istio-unit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rshriram commented:
/cherrypick release-1.3

istio-testing commented:
@rshriram: once the present PR merges, I will cherry-pick it on top of release-1.3 in a new PR and assign it to you.

<details>

In response to [this](https://github.com/istio/istio/pull/16691#issuecomment-526595841):

>/cherrypick release-1.3


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

hzxuzhonghu commented:
/hold

please fix the above comment

rshriram commented:
I distinctly remember seeing unit tests when I approved the PR (but they didn't pass). @maxbischoff please do add tests to make sure we dont regress.

istio-testing commented:
@maxbischoff: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-unit-tests-master | d1e82be570c6c460a32679628f689d8008db5895 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-unit-tests-master/6415) | `/test istio-unit-tests-master`
istio-lint-master | e02be1d77dd8bf19c1ce6136cb676c9d7376675c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-lint-master/3887) | `/test istio-lint-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@maxbischoff: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-lint-master | e02be1d77dd8bf19c1ce6136cb676c9d7376675c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-lint-master/3887) | `/test istio-lint-master`
istio-racetest-master | e02be1d77dd8bf19c1ce6136cb676c9d7376675c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-racetest-master/4606) | `/test istio-racetest-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

maxbischoff commented:
Thanks for the feedback, I changed it to use a map and added some unit tests. Is another release in 1.2 planned? If yes, can I cherry-pick it into release-1.2?

/retest

istio-testing commented:
@maxbischoff: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-racetest-master | 095b49b6e4688b19831bd6697b87a5b5a64fa86f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-racetest-master/4608) | `/test istio-racetest-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented:
/ok-to-test

istio-testing commented:
@maxbischoff: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-racetest-master | 095b49b6e4688b19831bd6697b87a5b5a64fa86f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-racetest-master/4610) | `/test istio-racetest-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@maxbischoff: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio-racetest-master | 095b49b6e4688b19831bd6697b87a5b5a64fa86f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/istio-racetest-master/4610) | `/test istio-racetest-master`
integ-mixer-k8s-presubmit-tests-master | 095b49b6e4688b19831bd6697b87a5b5a64fa86f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/integ-mixer-k8s-presubmit-tests-master/4794) | `/test integ-mixer-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

maxbischoff commented:
Test failures seem unrelated to my changes, or am I wrong?

/retest

istio-testing commented:
@rshriram: new pull request created: #16747

<details>

In response to [this](https://github.com/istio/istio/pull/16691#issuecomment-526595841):

>/cherrypick release-1.3


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@maxbischoff: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-mixer-k8s-presubmit-tests-master | 095b49b6e4688b19831bd6697b87a5b5a64fa86f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/16691/integ-mixer-k8s-presubmit-tests-master/4795) | `/test integ-mixer-k8s-presubmit-tests-master`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented:
There is a failure may be cause by this TestCreateGatewayHTTPFilterChainOpts
